Simplenote/src/main/java/com/automattic/simplenote/viewmodels/AddTagViewModel.kt

```diff
@@ -0,0 +1,70 @@
+package com.automattic.simplenote.viewmodels
+
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import com.automattic.simplenote.R
+import com.automattic.simplenote.repositories.TagsRepository
+
+class AddTagViewModel(private val tagsRepository: TagsRepository) : ViewModel() {
+    private val _uiState = MutableLiveData<UiState>()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _event = SingleLiveEvent<Event>()
+    val event: LiveData<Event> = _event
+
+    fun updateUiState(tagName: String) {
+        if (tagName.isEmpty()) {
+            _uiState.value = UiState(tagName, Status.ERROR, R.string.tag_error_empty)
+            return
+        }
+
+        if (tagName.contains(" ")) {
+            _uiState.value = UiState(tagName, Status.ERROR, R.string.tag_error_spaces)
+            return
+        }
+
+        if (!tagsRepository.isTagValid(tagName)) {
+            _uiState.value = UiState(tagName, Status.ERROR, R.string.tag_error_length)
+            return
+        }
+
+        if (!tagsRepository.isTagMissing(tagName)) {
+            _uiState.value = UiState(tagName, Status.ERROR, R.string.tag_error_exists)
+            return
+        }
+
+        _uiState.value = UiState(tagName, Status.VALID)
+    }
+
+    fun saveTag() {
+        val tagName = _uiState.value?.tagName ?: ""
+        _uiState.value = UiState(tagName, Status.SAVING)
+
+        val result = tagsRepository.saveTag(tagName)
+        _event.postValue(if (result) Event.FINISH else Event.SHOW_ERROR)
+    }
+
+    fun start() {
+        _uiState.value = UiState("", Status.STARTED)
+    }
+
+    fun close() {
+        _event.postValue(Event.CLOSE)
+    }
+
+    data class UiState(val tagName: String, val status: Status, val errorMsg: Int = -1)
+
+    enum class Status {
+        STARTED,
```

Comment:
This is looking really good! I was just wondering if the `STARTED` and the `SAVING` shouldn't be Events. What I'm thinking is that they are not really exclusive with tag/error states, right? If we were able to check whether the keyboard is already showing, I'd say that there should be a field on the `UiState` called something like `isKeyboardShowing`. However, if we want to keep the current behaviour, the `show keyboard` and `hide keyboard` are more events sent in `onCreate`, not really a state. 

If we do this change, we can remove the status field from the `UiState` and simplify it.

Reply:
Yeah, I think that change together with nullable of errorMessage can help remove the status field.

Reply:
Fixed in ee386d583f644a87bfff7074d9d38910947b955c.

Reply:
I don't understand why we need the `isKeyboardShowing` field on the UiState. Why can't we just hide keyboard on `FINISH` or `SHOW_ERROR` events? Is there a state I'm missing where we need to keep the information about the keyboard being hidden?

Reply:
Yes, makes sense and make the UiState simpler. Fixed in c1ba4baf52392ce9b648329474a66c6f6fb9859e.